File: app/public-category-rules.mjs
import {
  PUBLIC_CATEGORIES,
  PUBLIC_CATEGORY_ALIASES,
  PUBLIC_CATEGORY_GROUPS,
  PUBLIC_CATEGORY_LABEL_ALIASES,
  PUBLIC_CATEGORY_TAXONOMY,
  PUBLIC_EVENT_TYPE_LABELS,
} from "./public-category-taxonomy.generated.mjs";

const RULES = PUBLIC_CATEGORY_TAXONOMY.rules;
const FALLBACK_ID = PUBLIC_CATEGORY_TAXONOMY.fallback_category;
const CATEGORY_ORDER = new Map(
  (PUBLIC_CATEGORY_TAXONOMY.category_order || Object.keys(PUBLIC_CATEGORIES))
    .map((id, index) => [id, index]),
);
const TITLE_EVIDENCE_RULES = RULES.title_evidence.map((rule) => ({
  ...rule,
  regex: new RegExp(rule.pattern, "u"),
}));
const DESCRIPTION_EVIDENCE_RULES = RULES.description_evidence.map((rule) => ({
  ...rule,
  regex: new RegExp(rule.pattern, "u"),
}));
const SOURCE_TITLE_EVIDENCE_RULES = (RULES.source_title_evidence || []).map((rule) => ({
  ...rule,
  regex: new RegExp(rule.pattern, "u"),
}));
const SUMMER_PROGRAM_RE = new RegExp(RULES.summer_program_title_pattern, "u");
const SUMMER_REGISTRATION_RE = new RegExp(RULES.summer_registration_title_pattern, "u");
const SUMMER_PROGRAM_EVENT_TYPES = new Set(RULES.summer_program_event_types);
const SEMANTIC_NOISE_FIELDS = RULES.semantic_noise_fields || [];
const SEMANTIC_NOISE_PHRASES = RULES.semantic_noise_phrases || [];
const TAG_CATEGORY_WEIGHT = Number(RULES.tag_category_weight || 0);
const TAG_CATEGORY_ALIASES = RULES.tag_category_aliases || {};

export function foldPublicCategoryText(value) {
  return String(value || "")
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLocaleLowerCase("es")
    .replace(/[^a-z0-9]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function sourceCategory(event) {
  const source = event?.semantics?.source_category
    || event?.primary_category
    || event?.categories?.[0]
    || null;
  const label = String(source?.label || "").trim();
  const id = String(source?.id || foldPublicCategoryText(label).replace(/\s+/g, "-"))
    .trim()
    .toLocaleLowerCase("es");
  return { id, label };
}

export function canonicalPublicCategory(category) {
  const raw = typeof category === "string" ? { id: category, label: "" } : (category || {});
  const label = String(raw?.label || "").trim();
  const id = String(raw?.id || foldPublicCategoryText(label).replace(/\s+/g, "-"))
    .trim()
    .toLocaleLowerCase("es");
  const labelKey = foldPublicCategoryText(label);
  const canonicalId = PUBLIC_CATEGORY_ALIASES[id]
    || PUBLIC_CATEGORY_LABEL_ALIASES[labelKey]
    || (PUBLIC_CATEGORIES[id] ? id : null);
  if (!canonicalId) return id || label ? { id, label: label || id } : null;
  return { id: canonicalId, label: PUBLIC_CATEGORIES[canonicalId].label };
}

export function canonicalPublicCategoryId(category) {
  return canonicalPublicCategory(category)?.id || null;
}

export function isPublicCategoryInGroup(category, groupName) {
  const id = canonicalPublicCategoryId(category);
  return Boolean(id && (PUBLIC_CATEGORY_GROUPS[groupName] || []).includes(id));
}

export function publicCategorySymbol(category) {
  const id = canonicalPublicCategoryId(category);
  return id && PUBLIC_CATEGORIES[id]?.symbol
    ? PUBLIC_CATEGORIES[id].symbol
    : PUBLIC_CATEGORIES[FALLBACK_ID].symbol;
}

export function publicEventTypeLabel(eventType) {
  return PUBLIC_EVENT_TYPE_LABELS[String(eventType || "")] || null;
}

function category(id) {
  const resolvedId = PUBLIC_CATEGORIES[id] ? id : FALLBACK_ID;
  return { id: resolvedId, label: PUBLIC_CATEGORIES[resolvedId].label };
}

function isThematicCategory(id) {
  return Boolean(id && PUBLIC_CATEGORIES[id]?.thematic === true);
}

function isSummerProgram(event) {
  if (!SUMMER_PROGRAM_EVENT_TYPES.has(String(event?.event_type || ""))) return false;
  const title = foldPublicCategoryText(event?.title);
  return SUMMER_PROGRAM_RE.test(title) || SUMMER_REGISTRATION_RE.test(title);
}

function scalarNoiseValues(value) {
  if (typeof value === "string") return [value];
  if (!value || typeof value !== "object" || Array.isArray(value)) return [];
  const values = ["name", "label", "title", "address", "city", "venue"]
    .map((key) => value[key])
    .filter(Boolean)
    .map(String);
  const venue = foldPublicCategoryText(value.venue);
  const city = foldPublicCategoryText(value.city);
  if (venue && city && venue.endsWith(` ${city}`)) {
    const shortVenue = venue.slice(0, -(city.length + 1)).trim();
    if (shortVenue.length >= 4) values.push(shortVenue);
  }
  return values;
}

function semanticNoiseValues(event) {
  const values = [];
  for (const field of SEMANTIC_NOISE_FIELDS) {
    values.push(...scalarNoiseValues(event?.[field]));
  }
  values.push(...scalarNoiseValues(event?.source?.name));
  return [...new Set(values
    .map(foldPublicCategoryText)
    .filter((value) => value.length >= 4))]
    .sort((a, b) => b.length - a.length);
}

function escapedRegex(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function stripSemanticNoise(text, event) {
  let cleaned = foldPublicCategoryText(text);
  for (const phrase of SEMANTIC_NOISE_PHRASES) {
    const token = foldPublicCategoryText(phrase);
    if (token) cleaned = cleaned.replace(new RegExp(`\\b${escapedRegex(token)}\\b`, "gu"), " ");
  }
  for (const token of semanticNoiseValues(event)) {
    cleaned = cleaned.replace(new RegExp(`\\b${escapedRegex(token)}\\b`, "gu"), " ");
  }
  return cleaned.replace(/\s+/g, " ").trim();
}

function descriptionEvidenceText(event) {
  return stripSemanticNoise([
    event?.semantics?.category_evidence_text,
    event?.description,
    ...(event?.tags || []),
  ].filter(Boolean).join(" "), event);
}

function addEvidence(scores, evidence, categoryId, weight, kind, value) {
  if (!isThematicCategory(categoryId) || !weight) return;
  scores.set(categoryId, (scores.get(categoryId) || 0) + weight);
  evidence.push({ category: categoryId, weight, kind, value });
}

function addRuleEvidence(scores, evidence, text, rules, kind) {
  if (!text) return;
  for (const rule of rules) {
    if (rule.regex.test(text)) {
      addEvidence(scores, evidence, rule.category, Number(rule.weight || 0), kind, rule.pattern);
    }
  }
}

function addSourceTitleEvidence(scores, evidence, event) {
  const title = foldPublicCategoryText(event?.title);
  const sourceId = String(event?.source_id || "").trim();
  if (!title || !sourceId) return;
  for (const rule of SOURCE_TITLE_EVIDENCE_RULES) {
    if (String(rule.source_id || "") !== sourceId) continue;
    if (!rule.regex.test(title)) continue;
    addEvidence(
      scores,
      evidence,
      rule.category,
      Number(rule.weight || 0),
      "source_title",
      rule.reason || rule.pattern,
    );
  }
}

function addTagCategoryEvidence(scores, evidence, event) {
  if (!TAG_CATEGORY_WEIGHT) return;
  const seen = new Set();
  for (const rawTag of event?.tags || []) {
    const tag = foldPublicCategoryText(rawTag);
    const categoryId = TAG_CATEGORY_ALIASES[tag];
    const key = `${categoryId || ""}|${tag}`;
    if (!categoryId || seen.has(key)) continue;
    seen.add(key);
    addEvidence(scores, evidence, categoryId, TAG_CATEGORY_WEIGHT, "source_tag", rawTag);
  }
}

function confidenceForScore(score) {
  if (score >= 120) return "high";
  if (score >= 70) return "medium";
  return score >= Number(RULES.minimum_score || 1) ? "low" : "unclassified";
}

function rankedCandidates(scores, evidence) {
  return [...scores.entries()]
    .filter(([, score]) => score > 0)
    .sort((a, b) => (
      b[1] - a[1]
      || (CATEGORY_ORDER.get(a[0]) ?? Number.MAX_SAFE_INTEGER)
        - (CATEGORY_ORDER.get(b[0]) ?? Number.MAX_SAFE_INTEGER)
    ))
    .map(([categoryId, score]) => ({
      category: category(categoryId),
      score,
      confidence: confidenceForScore(score),
      evidence: evidence.filter((item) => item.category === categoryId),
    }));
}

export function classifyPublicCategory(event) {
  const scores = new Map();
  const evidence = [];
  const source = sourceCategory(event);
  const canonicalSource = canonicalPublicCategory(source);
  const recoveryHint = canonicalPublicCategory(event?.editorial?.category_recovery_hint || null);

  if (isSummerProgram(event)) {
    addEvidence(scores, evidence, "cursos-talleres-campus", 180, "event_type", "summer_program");
  }

  if (canonicalSource && isThematicCategory(canonicalSource.id)) {
    addEvidence(
      scores,
      evidence,
      canonicalSource.id,
      Number(RULES.source_category_weight || 0),
      "source_category",
      source.id || source.label,
    );
  }

  if (recoveryHint && isThematicCategory(recoveryHint.id)) {
    addEvidence(
      scores,
      evidence,
      recoveryHint.id,
      Number(RULES.recovery_hint_weight || RULES.source_category_weight || 0),
      "recovery_hint",
      String(event?.editorial?.category_recovery_hint || ""),
    );
  }

  // Generic venue/organizer/source-name text is intentionally excluded.
  // Verified sparse-title exceptions remain declarative in source_title_evidence.
  addTagCategoryEvidence(scores, evidence, event);
  addSourceTitleEvidence(scores, evidence, event);
  addRuleEvidence(
    scores,
    evidence,
    stripSemanticNoise(event?.title, event),
    TITLE_EVIDENCE_RULES,
    "title",
  );
  addRuleEvidence(
    scores,
    evidence,
    descriptionEvidenceText(event),
    DESCRIPTION_EVIDENCE_RULES,
    "description",
  );

  const candidates = rankedCandidates(scores, evidence);
  const minimumScore = Number(RULES.minimum_score || 1);
  const winner = candidates.find((candidate) => candidate.score >= minimumScore) || null;

  if (!winner) {
    return {
      category: category(FALLBACK_ID),
      confidence: "unclassified",
      score: candidates[0]?.score || 0,
      evidence,
      source_category: source,
      candidates,
    };
  }

  return {
    category: winner.category,
    confidence: winner.confidence,
    score: winner.score,
    evidence,
    source_category: source,
    candidates,
  };
}

export function resolvePublicCategory(event) {
  return classifyPublicCategory(event).category;
}

File: scripts/public_category_rules.py
from __future__ import annotations

import json
import re
import unicodedata
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
TAXONOMY_PATH = ROOT / "shared" / "public-category-taxonomy.json"
TAXONOMY = json.loads(TAXONOMY_PATH.read_text(encoding="utf-8"))
CATEGORIES: dict[str, dict[str, Any]] = TAXONOMY["categories"]
ALIASES: dict[str, str] = TAXONOMY["aliases"]
LABEL_ALIASES: dict[str, str] = TAXONOMY["label_aliases"]
GROUPS: dict[str, list[str]] = TAXONOMY["groups"]
RULES = TAXONOMY["rules"]
FALLBACK_ID = TAXONOMY["fallback_category"]
CATEGORY_ORDER = {
    category_id: index
    for index, category_id in enumerate(TAXONOMY.get("category_order", CATEGORIES.keys()))
}
SUMMER_PROGRAM_EVENT_TYPES = set(RULES["summer_program_event_types"])
SUMMER_PROGRAM_RE = re.compile(RULES["summer_program_title_pattern"])
SUMMER_REGISTRATION_RE = re.compile(RULES["summer_registration_title_pattern"])
TITLE_EVIDENCE_RULES = [
    (rule["category"], int(rule["weight"]), re.compile(rule["pattern"]))
    for rule in RULES["title_evidence"]
]
DESCRIPTION_EVIDENCE_RULES = [
    (rule["category"], int(rule["weight"]), re.compile(rule["pattern"]))
    for rule in RULES["description_evidence"]
]
SOURCE_EVIDENCE_RULES = [
    (rule["category"], int(rule["weight"]), re.compile(rule["pattern"]))
    for rule in RULES.get("source_evidence", [])
]
SOURCE_TITLE_EVIDENCE_RULES = [
    {**rule, "weight": int(rule["weight"]), "regex": re.compile(rule["pattern"])}
    for rule in RULES.get("source_title_evidence", [])
]
SEMANTIC_NOISE_FIELDS = tuple(RULES.get("semantic_noise_fields", []))
SEMANTIC_NOISE_PHRASES = tuple(RULES.get("semantic_noise_phrases", []))


def fold(value: Any) -> str:
    text = unicodedata.normalize("NFD", str(value or ""))
    text = "".join(char for char in text if unicodedata.category(char) != "Mn")
    text = text.casefold()
    text = re.sub(r"[^a-z0-9]+", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def source_category(event: dict[str, Any]) -> dict[str, str]:
    source = (
        (event.get("semantics") or {}).get("source_category")
        or event.get("primary_category")
        or (event.get("categories") or [None])[0]
        or {}
    )
    label = str(source.get("label") or "").strip()
    category_id = str(source.get("id") or re.sub(r"\s+", "-", fold(label))).strip().casefold()
    return {"id": category_id, "label": label}


def canonical_public_category(category: dict[str, Any] | str | None) -> dict[str, str] | None:
    raw = {"id": category, "label": ""} if isinstance(category, str) else (category or {})
    label = str(raw.get("label") or "").strip()
    category_id = str(raw.get("id") or re.sub(r"\s+", "-", fold(label))).strip().casefold()
    canonical_id = ALIASES.get(category_id) or LABEL_ALIASES.get(fold(label))
    if not canonical_id and category_id in CATEGORIES:
        canonical_id = category_id
    if canonical_id:
        return {"id": canonical_id, "label": CATEGORIES[canonical_id]["label"]}
    if category_id or label:
        return {"id": category_id, "label": label or category_id}
    return None


def canonical_public_category_id(category: dict[str, Any] | str | None) -> str | None:
    resolved = canonical_public_category(category)
    return resolved.get("id") if resolved else None


def is_public_category_in_group(category: dict[str, Any] | str | None, group_name: str) -> bool:
    category_id = canonical_public_category_id(category)
    return bool(category_id and category_id in GROUPS.get(group_name, []))


def category(category_id: str) -> dict[str, str]:
    resolved_id = category_id if category_id in CATEGORIES else FALLBACK_ID
    return {"id": resolved_id, "label": CATEGORIES[resolved_id]["label"]}


def is_thematic_category(category_id: str | None) -> bool:
    return bool(category_id and CATEGORIES.get(category_id, {}).get("thematic") is True)


def is_summer_program(event: dict[str, Any]) -> bool:
    if str(event.get("event_type") or "") not in SUMMER_PROGRAM_EVENT_TYPES:
        return False
    title = fold(event.get("title"))
    return bool(SUMMER_PROGRAM_RE.search(title) or SUMMER_REGISTRATION_RE.search(title))


def _scalar_noise_values(value: Any) -> list[str]:
    if isinstance(value, str):
        return [value]
    if isinstance(value, dict):
        return [
            str(value.get(key) or "")
            for key in ("name", "label", "title", "address", "city", "venue")
            if value.get(key)
        ]
    return []


def _semantic_noise_values(event: dict[str, Any]) -> list[str]:
    values: list[str] = []
    for field in SEMANTIC_NOISE_FIELDS:
        values.extend(_scalar_noise_values(event.get(field)))
    source = event.get("source") if isinstance(event.get("source"), dict) else {}
    values.extend(_scalar_noise_values(source.get("name")))
    return [fold(value) for value in values if len(fold(value)) >= 4]


def _strip_semantic_noise(text: str, event: dict[str, Any]) -> str:
    cleaned = fold(text)
    for phrase in SEMANTIC_NOISE_PHRASES:
        token = fold(phrase)
        if token:
            cleaned = re.sub(rf"(?<!\w){re.escape(token)}(?!\w)", " ", cleaned)
    for token in sorted(set(_semantic_noise_values(event)), key=len, reverse=True):
        cleaned = re.sub(rf"(?<!\w){re.escape(token)}(?!\w)", " ", cleaned)
    return re.sub(r"\s+", " ", cleaned).strip()


def description_evidence_text(event: dict[str, Any]) -> str:
    values = [event.get("description"), *(event.get("tags") or [])]
    return _strip_semantic_noise(
        " ".join(str(value) for value in values if value),
        event,
    )


def source_evidence_text(event: dict[str, Any]) -> str:
    # Kept as a compatibility helper. Generic source/organizer text is no longer
    # semantic category evidence; source-specific evidence must be declarative.
    return ""


def _add_evidence(
    scores: dict[str, int],
    evidence: list[dict[str, Any]],
    category_id: str,
    weight: int,
    kind: str,
    value: Any,
) -> None:
    if not is_thematic_category(category_id) or not weight:
        return
    scores[category_id] = scores.get(category_id, 0) + weight
    evidence.append(
        {"category": category_id, "weight": weight, "kind": kind, "value": value}
    )


def _add_rule_evidence(
    scores: dict[str, int],
    evidence: list[dict[str, Any]],
    text: str,
    rules: list[tuple[str, int, re.Pattern[str]]],
    kind: str,
) -> None:
    if not text:
        return
    for category_id, weight, pattern in rules:
        if pattern.search(text):
            _add_evidence(scores, evidence, category_id, weight, kind, pattern.pattern)


def _add_source_title_evidence(
    scores: dict[str, int],
    evidence: list[dict[str, Any]],
    event: dict[str, Any],
) -> None:
    title = fold(event.get("title"))
    source_id = str(event.get("source_id") or "").strip()
    if not title or not source_id:
        return
    for rule in SOURCE_TITLE_EVIDENCE_RULES:
        if str(rule.get("source_id") or "") != source_id:
            continue
        if rule["regex"].search(title):
            _add_evidence(
                scores,
                evidence,
                str(rule["category"]),
                int(rule["weight"]),
                "source_title",
                rule.get("reason") or rule["pattern"],
            )


def _confidence_for_score(score: int) -> str:
    if score >= 120:
        return "high"
    if score >= 70:
        return "medium"
    if score >= int(RULES.get("minimum_score", 1)):
        return "low"
    return "unclassified"


def _ranked_candidates(
    scores: dict[str, int], evidence: list[dict[str, Any]]
) -> list[dict[str, Any]]:
    ranked = sorted(
        ((category_id, score) for category_id, score in scores.items() if score > 0),
        key=lambda item: (-item[1], CATEGORY_ORDER.get(item[0], 10_000)),
    )
    return [
        {
            "category": category(category_id),
            "score": score,
            "confidence": _confidence_for_score(score),
            "evidence": [item for item in evidence if item.get("category") == category_id],
        }
        for category_id, score in ranked
    ]


def classify_public_category(event: dict[str, Any]) -> dict[str, Any]:
    scores: dict[str, int] = {}
    evidence: list[dict[str, Any]] = []
    source = source_category(event)
    canonical_source = canonical_public_category(source)
    recovery_hint = canonical_public_category(
        (event.get("editorial") or {}).get("category_recovery_hint")
    )

    if is_summer_program(event):
        _add_evidence(
            scores, evidence, "cursos-talleres-campus", 180, "event_type", "summer_program"
        )

    if canonical_source and is_thematic_category(canonical_source.get("id")):
        _add_evidence(
            scores,
            evidence,
            canonical_source["id"],
            int(RULES.get("source_category_weight", 0)),
            "source_category",
            source.get("id") or source.get("label"),
        )

    if recovery_hint and is_thematic_category(recovery_hint.get("id")):
        _add_evidence(
            scores,
            evidence,
            recovery_hint["id"],
            int(RULES.get("recovery_hint_weight", RULES.get("source_category_weight", 0))),
            "recovery_hint",
            (event.get("editorial") or {}).get("category_recovery_hint"),
        )

    # Generic venue/organizer/source-name text is intentionally excluded.
    # Verified sparse-title exceptions remain declarative in source_title_evidence.
    _add_source_title_evidence(scores, evidence, event)
    _add_rule_evidence(
        scores, evidence, fold(event.get("title")), TITLE_EVIDENCE_RULES, "title"
    )
    _add_rule_evidence(
        scores, evidence, description_evidence_text(event), DESCRIPTION_EVIDENCE_RULES, "description"
    )

    candidates = _ranked_candidates(scores, evidence)
    minimum_score = int(RULES.get("minimum_score", 1))
    winner = next(
        (candidate for candidate in candidates if candidate["score"] >= minimum_score),
        None,
    )

    if not winner:
        return {
            "category": category(FALLBACK_ID),
            "confidence": "unclassified",
            "score": candidates[0]["score"] if candidates else 0,
            "evidence": evidence,
            "source_category": source,
            "candidates": candidates,
        }

    return {
        "category": winner["category"],
        "confidence": winner["confidence"],
        "score": winner["score"],
        "evidence": evidence,
        "source_category": source,
        "candidates": candidates,
    }


def resolve_public_category(event: dict[str, Any]) -> dict[str, str]:
    return dict(classify_public_category(event)["category"])


def public_category_text(event: dict[str, Any]) -> str:
    return resolve_public_category(event)["label"]
